File: src/main/java/no/fintlabs/response/ResponseFintEventEntity.java
package no.fintlabs.response;

import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import lombok.*;

@Entity
@Getter
@NoArgsConstructor
public class ResponseFintEventEntity {

    @Id
    public String corrId;
    public String orgId;
    public String adapterId;
    public String topic;
    public long handledAt;
    public boolean failed;
    public String errorMessage;
    public String rejectReason;
    public boolean conflicted;
    public boolean rejected;
    public String conflictReason;
    public String syncPageEntryIdentifier;
    public String syncPageEntryResource;

    public ResponseFintEventEntity(String corrId, String orgId, String adapterId, String topic, long handledAt, boolean failed, String errorMessage, String rejectReason, boolean conflicted, boolean rejected, String conflictReason, String syncPageEntryIdentifier, String syncPageEntryResource) {
        this.corrId = corrId;
        this.orgId = orgId;
        this.adapterId = adapterId;
        this.topic = topic;
        this.handledAt = handledAt;
        this.failed = failed;
        this.errorMessage = errorMessage;
        this.rejectReason = rejectReason;
        this.conflicted = conflicted;
        this.rejected = rejected;
        this.conflictReason = conflictReason;
        this.syncPageEntryIdentifier = syncPageEntryIdentifier;
        this.syncPageEntryResource = syncPageEntryResource;
    }
}

File: src/main/kotlin/no/fintlabs/event/response/ResponseFintEventEntity.kt
package no.fintlabs.event.response

import jakarta.persistence.Entity
import jakarta.persistence.Id
import no.fintlabs.adapter.models.event.ResponseFintEvent
import no.fintlabs.adapter.models.sync.SyncPageEntry

@Entity
class ResponseFintEventEntity(
    @Id var corrId: String?,
    var orgId: String?,
    var adapterId: String?,
    var topic: String?,
    var handledAt: Long,
    var failed: Boolean,
    var errorMessage: String?,
    var rejectReason: String?,
    var conflicted: Boolean,
    var rejected: Boolean,
    var conflictReason: String?,
    var syncPageEntryIdentifier: String?,
    var syncPageEntryResource: String?
) {
    constructor() : this(null, null, null, null, 0L, false, null, null, false, false, null, null, null)

    fun toResponseEvent(): ResponseFintEvent {
        return ResponseFintEvent.builder()
            .corrId(corrId)
            .orgId(orgId)
            .adapterId(adapterId)
            .handledAt(handledAt)
            .failed(failed)
            .errorMessage(errorMessage)
            .rejected(rejected)
            .rejectReason(rejectReason)
            .conflicted(conflicted)
            .conflictReason(conflictReason)
            .value(SyncPageEntry.of(syncPageEntryIdentifier, syncPageEntryResource))
            .build()
    }

    companion object {
        fun fromResponseEvent(responseEvent: ResponseFintEvent, topic: String): ResponseFintEventEntity {
            return ResponseFintEventEntity(
                responseEvent.corrId,
                responseEvent.orgId,
                responseEvent.adapterId,
                topic,
                responseEvent.handledAt,
                responseEvent.isFailed,
                responseEvent.errorMessage,
                responseEvent.rejectReason,
                responseEvent.isConflicted,
                responseEvent.isRejected,
                responseEvent.conflictReason,
                responseEvent.value?.identifier ?: "[NULL]",
                ""
            )
        }
    }
}
